### PR TITLE
Added Custom File-Not-Found page for Tabs

### LINKED ISSUE #1073: Remember current tab from previous user session
#1048 introduced the feature to reopen tabs on start-up. However, on many browsers (Firefox, Chrome), the current tab is also being kept track of when restoring user sessions. It would be good UX to have this feature.

One thing to discuss is: that we are not currently reopening invalid Zim tabs, empty tabs, settings tabs, etc. Users could be on one of those tabs from the previous session, including the library tab. How should we define the behavior of our tab reopenings?

### PR DESCRIPTION
**Pull request by ShaopengLin**
Fix #1073
Follow-up from #1120.

Changes:
- Custom file-not-found page when tabs have Urls that point to non-existing Zim files.
- Now displays the bookmarks of the non-existing Zim files. Previously they were marked as invalid and not added.

### CODE REVIEW

**Comment by ShaopengLin** [comment]
Let's review this after #1120 is merged. The reopening functionality will be available after we rebase this.

File: src/readinglistbar.cpp
Comment (kelson42) [reply src/readinglistbar.cpp:52]: no

Comment (ShaopengLin) [reply src/readinglistbar.cpp:52]: hmm, somehow I have a bunch of entries that don't have zimIds. It might be a bug but I am not sure how to reproduce.

File: src/readinglistbar.cpp
Comment (ShaopengLin) [reply src/readinglistbar.cpp:114]: No, we were only adding bookmarks that are valid in setupList. The getBookmarks(true) also only looks at valid bookmarks. The order were the same so we didn't have a bug before.

**Comment by ShaopengLin** [comment]
> Why were the changes related to a more informative File-Not-Found error page (reporting the ZIM path) lost?

@veloman-yunkan sorry I am not exactly sure what this means? I thought I removed everything about zim path and only did a custom error page.

**Comment by veloman-yunkan** [comment]
> > Why were the changes related to a more informative File-Not-Found error page (reporting the ZIM path) lost?
> 
> @veloman-yunkan sorry I am not exactly sure what this means? I thought I removed everything about zim path and only did a custom error page.

I thought that this PR should contain all changes that were removed from #1120 when we [decided to split it into two parts](https://github.com/kiwix/kiwix-desktop/pull/1120#issuecomment-2171427504).

**Comment by ShaopengLin** [comment]
I was looking at kelson's comment and thought that you meant we are going go for that. Should I still add the path? https://github.com/kiwix/kiwix-desktop/pull/1120#issuecomment-2169248914

**Comment by veloman-yunkan** [comment]
> If we were to add the missing paths we will have to go back to that original complicated solution.

And that is exactly why I chose to split the PR into two parts - so that the complicated solution doesn't block merging the fix for the original issue while we keep working around problems that crop up while we pursue a secondary improvement.

File: src/readinglistbar.cpp
Comment (ShaopengLin) [reply src/readinglistbar.cpp]: Done

File: resources/i18n/en.json
Comment (ShaopengLin) [reply resources/i18n/en.json]: Done

File: resources/i18n/en.json
Comment (ShaopengLin) [reply resources/i18n/en.json]: Done

File: src/readinglistbar.cpp
Comment (veloman-yunkan) [reply src/readinglistbar.cpp:53]: Let's try **not** to hide any invalid bookmarks. But we can display them in a special way.

Note that if you add all bookmarks to the reading list, you won't need the extra `idx` parameter in `ReadingListBar::addItem()`

File: src/tabbar.h
Comment (veloman-yunkan) [inline src/tabbar.h:28]: When you were referring to the complicated solution I thought that you meant the last version (that had some issues because of directory monitoring) before we decided to split the PR. Sorry.

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]
I think that we should again split this PR into two:

1. Improvement of the missing ZIM file error page. The main challenge with this one will be to resolve the issues that we faced just before the previous split (related to the side effects of directory monitoring).
2. Listing all bookmarks (including invalid ones)

File: src/tabbar.h
Comment (ShaopengLin) [reply src/tabbar.h:28]: Since we cannot change the behavior of directory monitoring, we can:
1. simply save all the books that we have seen in updateFromDir, no matter if we will be removing them.
2. only remove a saved book when all traces such as opened tab and bookmarks of that book is removed.

**Comment by ShaopengLin** [comment]
@veloman-yunkan This solution is a lot simpler than the last one. If you have any other suggestions I am open to try.

File: src/kiwixapp.cpp
Comment (veloman-yunkan) [inline src/kiwixapp.cpp:626]: To me something seems wrong about saving this information in the session. Please give me some time to come up with a better solution or admit that none exists.

**Comment by kelson42** [comment]
@ShaopengLin Any feedback?

**Comment by ShaopengLin** [comment]
> To me something seems wrong about saving this information in the session. Please give me some time to come up with a better solution or admit that none exists.

We are a little stuck on how to properly design the solution to this problem completely without changing the behaviour of the directory monitoring. I believe veloman is busy with the other issue and hasn't touched on this yet.

@veloman-yunkan @kelson42  This has been stuck for a while. We don't have an easy path to save those information otherwise, since the directory monitoring completely removes the record of zims if it disappears.

I believe a clean solution would likely need us to either create a new file like library.xml to retain the missing zim entries, or change libkiwix to save this information.

**Comment by ShaopengLin** [comment]
@veloman-yunkan This has been stuck for too long. If storing this information in settings or a file doesn't make sense, we can store it in the URL paths. What do you think?

Otherwise, we should simply do what Kelson said, have a simple error message with Zim Id, and then move on to have the bookmark portion finished. Having the bookmark feature I believe is an equally important user experience.

**Comment by kelson42** [comment]
@ShaopengLin I would like to intervene here as things don't move well and maybe not even in the right direction.

First of all, to me, all the work around bookmarks in that PR is unclear: what is tried to be achieved and what is the chosen approach. One reason is that there is no proper issue describing the problem and therefore no formal validation about both problem and solution approach.

Regarding the topic of "invalid" bookmarks, to me this is not a problem because:
* Bookmarks refering to a ZIM which is not in the local library should never be displayed
* All the bookmark handling should be done in the libkiwix (not in Kiwix Desktop)

For all these reasons, please:
* Keep only the first part of your PR here (about missing page error)
* Open an issue describing clearly the problem around the bookmark so we can discuss it independently of any PR

**Comment by veloman-yunkan** [comment]
> @veloman-yunkan This has been stuck for too long. If storing this information in settings or a file doesn't make sense, we can store it in the URL paths. What do you think?

@ShaopengLin I still think that the approach outlined in https://github.com/kiwix/kiwix-desktop/pull/1120#issuecomment-2162262938 is the best. A proper implementation of it would require moving the directory monitoring functionality from `Library` to `ContentManager` - the former doesn't have any knowledge of a book being opened in a tab while the latter does. Then `ContentManager` will be in charge of deciding how to handle a disappeared ZIM file - if it is of special interest to the user (e.g. was open in a tab, was bookmarked, etc) then it will be preserved in the library but marked as missing, otherwise it will be simply dropped from the library.

If you think it's too complex or get stuck while pursuing that solution please let me know and I will implement it.

**Comment by ShaopengLin** [comment]
@veloman-yunkan Kelson has mentioned this before:
> There are many valid reasons why the directory monitoring can remove a file from the library and all of them are external to Kiwix. We have no control on this. The rest must follow.

Doesn't this violate our solution to keep the book in library.xml? This was the reason I was not going with the solution from [#1120 (comment)](https://github.com/kiwix/kiwix-desktop/pull/1120#issuecomment-2162262938) for this PR in the first place.

**Comment by veloman-yunkan** [comment]
> I propose that I implement the said change in library monitoring as a separate PR whereupon you can rebase and test your PR on top of my branch.

Done in #1149. You will have to enhance the function `ContentManager::handleDisappearedZimFile()` as described in https://github.com/kiwix/kiwix-desktop/pull/1132#issuecomment-2238540879.

**Comment by ShaopengLin** [comment]
@veloman-yunkan I have rebased and tested your change. Do you mean I simply push what I have (my change rebased onto yours.) to this PR?

**Comment by veloman-yunkan** [comment]
> @veloman-yunkan I have rebased and tested your change. Do you mean I simply push what I have (my change rebased onto yours.) to this PR?

Yes (and change the base branch to `directory_monitoring_improvements` by editing the PR so that only your changes show up here).

File: src/contentmanager.cpp
Comment (ShaopengLin) [inline src/contentmanager.cpp:910]: @veloman-yunkan I checked that the book's isPathValid boolean was already set when we retrieved it. Don't think we need to make any modifications to the library.

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]
It works! Nice!

Please separate the change addressing the subtlety with directory monitoring into a separate commit. Then this PR will be complete.

**Comment by ShaopengLin** [comment]
> Please separate the change addressing the subtlety with directory monitoring into a separate commit. Then this PR will be complete.

I assume you meant the setupDirectoryMonitoring subtlety

**Comment by veloman-yunkan** [comment]
> > Please separate the change addressing the subtlety with directory monitoring into a separate commit. Then this PR will be complete.
> 
> I assume you meant the setupDirectoryMonitoring subtlety

No, I meant everything related to directory monitoring. Suppose that you implement the more informative file-not-found error page with directory monitoring disabled. Then you find out that turning directory monitoring on leads to certain issues. I want those issues to be addressed in a separate commit (with a proper explanation).

**Comment by ShaopengLin** [comment]
@veloman-yunkan 
There are two situations where external modification to the library due to directory monitoring is problematic:

1. If the library file(s) are removed or modified, our code for file-not-found page can fault due to the library losing book information. This is, however already mitigated with the try/catch when we are unable to find the book.

2. Existing functions may rely on the previous monitoring behaviour to validate whether the zim file physically exist. The getValidBookById resolves this issue. This function retains the previous behaviour of getBookById, and should be used if we want to get a book that is physically present. Could you do a sanity check on the getBookById calls I left out or replaced to make sure I didn't make any misconception of the intentions of getBookById? Thanks!

**Review by veloman-yunkan — COMMENTED** [review]
> @veloman-yunkan There are two situations where external modification to the library due to directory monitoring is problematic:
> 
>  1. If the library file(s) are removed or modified, our code for file-not-found page can fault due to the library losing book information. This is, however already mitigated with the try/catch when we are unable to find the book.
> 
>  2. Existing functions may rely on the previous monitoring behaviour to validate whether the zim file physically exist. The getValidBookById resolves this issue. This function retains the previous behaviour of getBookById, and should be used if we want to get a book that is physically present. Could you do a sanity check on the getBookById calls I left out or replaced to make sure I didn't make any misconception of the intentions of getBookById? Thanks!

Let's not let this PR's scope drift. The purpose of this PR is to improve the information displayed in the file-not-found error page.

Existing functionality does *not* rely on previous behaviour of directory monitoring. Directory monitoring was added relatively recently and not all of its side effects were fully understood and addressed. Let's not go after potential issues that have been introduced due to directory monitoring in the past. Just make sure that your enhancement

1. works in the absence of directory monitoring in the first place and
2. also works when directory monitoring is enabled (changes targeting this part should come in a separate commit).

**Comment by kelson42** [comment]
#1149 has been merged, so if I get it right, after a rebase it should be possible tom omplete this PR!?

File: src/tabbar.h
Comment (veloman-yunkan) [inline src/tabbar.h:53]: This function should be introduced in the next commit.

File: src/urlschemehandler.cpp
Comment (veloman-yunkan) [inline src/urlschemehandler.cpp:192]: I think that this change is unrelated to directory monitoring - it can be merged into the previous commit.

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]
Almost there. Please fix the last two comments and also format your commit messages to meet the [commit message guidelines](https://stackoverflow.com/questions/2290016/git-commit-messages-50-72-formatting)

File: src/tabbar.h
Comment (ShaopengLin) [reply src/tabbar.h:53]: Done

File: src/urlschemehandler.cpp
Comment (ShaopengLin) [reply src/urlschemehandler.cpp]: Done

**Review by veloman-yunkan — APPROVED** [review]

**Pull request merged**
